File: flutter/ios/Classes/UnfydqryPlugin.swift
import Flutter
import UIKit
import UnifiedQuery

/// iOS side of the Flutter plugin.
///
/// Each open engine lives in ``engines`` keyed by an integer handle that is
/// returned to Dart on "open" and sent back on every subsequent call.
public class UnfydqryPlugin: NSObject, FlutterPlugin {

    private var engines: [Int: SearchEngine] = [:]
    private var nextHandle = 0

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "unfydqry/search",
            binaryMessenger: registrar.messenger()
        )
        let instance = UnfydqryPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        do {
            switch call.method {

            case "open":
                let dbPath = args["dbPath"] as! String
                let engine = try SearchEngine(dbPath: dbPath)
                let handle = nextHandle
                nextHandle += 1
                engines[handle] = engine
                result(handle)

            case "index":
                let engine = try requireEngine(args)
                try engine.index(id: int64(args["id"]!), text: args["text"] as! String)
                result(nil)

            case "remove":
                let engine = try requireEngine(args)
                try engine.remove(id: int64(args["id"]!))
                result(nil)

            case "search":
                let engine = try requireEngine(args)
                let hits = try engine.search(
                    query: args["query"] as! String,
                    limit: UInt32(args["limit"] as! Int)
                )
                result(hits.map { ["id": $0.id, "score": $0.score] })

            case "dispose":
                let handle = args["handle"] as! Int
                engines.removeValue(forKey: handle)
                result(nil)

            default:
                result(FlutterMethodNotImplemented)
            }
        } catch let error as SearchError {
            result(FlutterError(code: "SEARCH_ERROR", message: error.localizedDescription, details: nil))
        } catch {
            result(FlutterError(code: "PLUGIN_ERROR", message: error.localizedDescription, details: nil))
        }
    }

    private func requireEngine(_ args: [String: Any]) throws -> SearchEngine {
        let handle = args["handle"] as! Int
        guard let engine = engines[handle] else {
            throw PluginError.noEngine(handle)
        }
        return engine
    }

    // Flutter's standard message codec boxes numbers as NSNumber.
    private func int64(_ value: Any) -> Int64 {
        (value as! NSNumber).int64Value
    }
}

private enum PluginError: LocalizedError {
    case noEngine(Int)
    var errorDescription: String? {
        switch self {
        case .noEngine(let h): return "no engine for handle \(h)"
        }
    }
}

File: flutter/android/src/main/kotlin/unfydqry/flutter/UnfydqryPlugin.kt
package unfydqry.flutter

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import uniffi.unfydqry.SearchEngine
import uniffi.unfydqry.SearchException

/**
 * Android side of the Flutter plugin.
 *
 * Each open engine is kept in [engines] under an integer handle that is
 * returned to Dart on 'open' and passed back on every subsequent call.
 */
class UnfydqryPlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel
    private val engines = mutableMapOf<Int, SearchEngine>()
    private var nextHandle = 0

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "unfydqry/search")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        engines.values.forEach { it.close() }
        engines.clear()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        try {
            when (call.method) {
                "open" -> {
                    val dbPath = call.argument<String>("dbPath")!!
                    val handle = nextHandle++
                    engines[handle] = SearchEngine(dbPath)
                    result.success(handle)
                }

                "index" -> {
                    engine(call).index(
                        id = call.longArg("id"),
                        text = call.argument<String>("text")!!,
                    )
                    result.success(null)
                }

                "remove" -> {
                    engine(call).remove(id = call.longArg("id"))
                    result.success(null)
                }

                "search" -> {
                    val hits = engine(call).search(
                        query = call.argument<String>("query")!!,
                        limit = call.argument<Int>("limit")!!.toUInt(),
                    )
                    result.success(hits.map { mapOf("id" to it.id, "score" to it.score) })
                }

                "dispose" -> {
                    val handle = call.argument<Int>("handle")!!
                    engines.remove(handle)?.close()
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        } catch (e: SearchException) {
            result.error("SEARCH_ERROR", e.message, null)
        } catch (e: Exception) {
            result.error("PLUGIN_ERROR", e.message, null)
        }
    }

    private fun engine(call: MethodCall): SearchEngine =
        engines[call.argument<Int>("handle")!!]
            ?: error("no engine for handle ${call.argument<Int>("handle")}")

    // Flutter's method channel can deliver Dart int as Int or Long depending on value.
    private fun MethodCall.longArg(key: String): Long =
        (argument<Any>(key) as Number).toLong()
}
